pkg/distributor/forwarding/forwarding.go

```diff
@@ -96,185 +74,260 @@ func NewForwarder(reg prometheus.Registerer, cfg Config) Forwarder {
 		}),
 	}
+
+	f.workerWg.Add(f.cfg.RequestConcurrency)
+	for i := 0; i < f.cfg.RequestConcurrency; i++ {
+		go f.worker()
+	}
+
+	return f
 }
 
-func (r *forwarder) NewRequest(ctx context.Context, tenant string, rules validation.ForwardingRules) Request {
-	return &request{
-		ctx:    ctx,
-		client: &r.client, // http client should be re-used so open connections get re-used.
-		pools:  &r.pools,
+func (f *forwarder) Stop() {
+	close(f.reqCh)
+	f.workerWg.Wait()
+}
 
-		tsByEndpoint: make(map[string]*[]mimirpb.PreallocTimeseries),
+func (f *forwarder) Forward(ctx context.Context, rules validation.ForwardingRules, in []mimirpb.PreallocTimeseries) (TimeseriesCounts, []mimirpb.PreallocTimeseries, chan error) {
+	if !f.cfg.Enabled {
+		errCh := make(chan error)
+		close(errCh)
+		return TimeseriesCounts{}, in, errCh
+	}
 
-		rules:           rules,
-		timeout:         r.cfg.RequestTimeout,
-		propagateErrors: r.cfg.PropagateErrors,
+	notIngestedCounts, toIngest, tsByTargets := f.splitByTargets(in, rules)
+	defer f.pools.putTsByTargets(tsByTargets)
 
-		requests: r.requestsTotal,
-		errors:   r.errorsTotal,
-		samples:  r.samplesTotal,
-		latency:  r.requestLatencyHistogram,
+	var requestWg sync.WaitGroup
+	requestWg.Add(len(tsByTargets))
+	errCh := make(chan error, len(tsByTargets))
+	for endpoint, ts := range tsByTargets {
+		f.newRequest(ctx, endpoint, ts, &requestWg, errCh)
 	}
+
+	go func() {
+		requestWg.Wait()
+		close(errCh)
+	}()
+
+	return notIngestedCounts, toIngest, errCh
 }
 
-type request struct {
-	ctx    context.Context
-	client *http.Client
-	pools  *pools
+type tsWithSampleCount struct {
+	ts     []mimirpb.PreallocTimeseries
+	counts TimeseriesCounts
+}
 
-	tsByEndpoint map[string]*[]mimirpb.PreallocTimeseries
+type tsByTargets map[string]tsWithSampleCount
 
-	// The rules which define:
-	// - which metrics get forwarded
-	// - where the metrics get forwarded to
-	// - whether the forwarded metrics should also be ingested (sent to ingesters)
-	rules           validation.ForwardingRules
-	timeout         time.Duration
-	propagateErrors bool
+type TimeseriesCounts struct {
+	SampleCount   int
+	ExemplarCount int
+}
 
-	requests prometheus.Counter
-	errors   *prometheus.CounterVec
-	samples  prometheus.Counter
-	latency  prometheus.Histogram
+func (t *TimeseriesCounts) count(ts mimirpb.PreallocTimeseries) {
+	t.SampleCount += len(ts.TimeSeries.Samples)
+	t.ExemplarCount += len(ts.TimeSeries.Exemplars)
 }
 
-func (r *request) Add(sample mimirpb.PreallocTimeseries) bool {
-	metric, err := extract.UnsafeMetricNameFromLabelAdapters(sample.Labels)
-	if err != nil {
-		// The only possible error is due to no metric name being defined, in which case we won't forward the sample
-		// to anywhere but we should still send it to the Ingesters.
-		return true
-	}
+// forwardingEndpointAssigner returns a func which takes a timeseries and assigns it endpoints to which it should be forwarded.
```

Comment:
Outdated comment.

Reply:
thx! 
fixed in 1f86587e23e0b3fea2bc0137c3514dc7d42021e9